### DISCUSSION TITLE
Tailwind CSS not working in WSL

### DISCUSSION THREAD

**Discussion by geeekgod** [tailwindlabs/tailwindcss#10302] [Help] [answered]
**What version of Tailwind CSS are you using?**

I'm using the gem [tailwindcss-rails](https://github.com/rails/tailwindcss-rails) which has tailwindcss [v3.2.4](https://github.com/tailwindlabs/tailwindcss/releases/tag/v3.2.4)

**What build tool (or framework if it abstracts the build tool) are you using?**

I'm not using any build tool since I'm using Rails 7 with import maps.

**What version of Node.js are you using?**

v16.17.0

**What browser are you using?**

Firefox

**What operating system are you using?**

WSL Ubuntu 20.04

**Reproduction URL**
[tailwind-rails-bug](https://github.com/thisisrishabh22/tailwind-rails-bug)

**Describe your issue**

I'm trying to use tailwindcss in my rails app using the official [documentation](https://tailwindcss.com/docs/guides/ruby-on-rails), but unable to run the app since tailwindcss is not working in my WSL system. I opened an issue [Help with tailwindcss-rails setup](https://github.com/rails/tailwindcss-rails/issues/235), the team informed me to open a bug report since the binary file is causing the problem.

**Comment by thecrypticace** [comment]
Hey! You are getting an error because you are on a 32-bit system and installing / running a 64-bit binary. We do not compile binaries for 32-bit linux because `pkg` (the tool we use to package node + tailwindcss into a standalone binary) doesn't support it. I don't think Node.js itself has **_official_** builds for 32-bit linux which is probably why it's not supported by pkg.

Unfortunately, I don't have a workaround other than using a 64-bit system because building a binary for 32-bit linux isn't possible currently.

**Comment by geeekgod** [comment]
Hi, @thecrypticace I'm actually using a 64-bit system.
```shell
Linux LAPTOP-O1H43SIF 4.4.0-22621-Microsoft #608-Microsoft Fri Jan 01 08:00:00 PST 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by thecrypticace** [comment]
@thisisrishabh22 

What about your windows installation? Is it 32-bit and emulating 64-bit linux? The `ELF…` message indicates that the system you are on does not understand the ELF 64-bit executable format. I'm running the same version of Ubuntu in WSL2 on my Windows machine and things run fine.

Can you provide output from the following two commands in WSL?

- `file ./tailwindcss-linux-x64` (you'll need to be in the appropriate folder)
- `file /bin/ls`

**Comment by geeekgod** [comment]
@thecrypticace 

I'm using a windows 11 machine with a `64-bit operating system, x64-based processor`

Here are the output for the commands:

- `file ./tailwindcss-linux-x64`
```shell
./tailwindcss-linux-x64: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, stripped
```

- `file /bin/ls`
```shell
/bin/ls: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=2f15ad836be3339dec0e2e6a3c637e08e48aacbd, for GNU/Linux 3.2.0, stripped
```

**Comment by thecrypticace** [comment]
@thisisrishabh22 

What happens if you download the binary directly from our releases page? Can you run that one?

```
wget https://github.com/tailwindlabs/tailwindcss/releases/download/v3.2.4/tailwindcss-linux-x64
chmod +x ./tailwindcss-linux-x64
./tailwindcss-linux-x64 --help
```

**Comment by geeekgod** [comment]
I've tried running the binary file directly from the releases page as well.
You can check the output [here](https://github.com/rails/tailwindcss-rails/issues/235#issuecomment-1359959810)

**Comment by thecrypticace** [comment]
The only difference that I can discern between my WSL installation and yours is the kernel version. I'm apparently running kernel 4.19 whereas you are on 4.4. I don't know if there's maybe a bug there. But seems unlikely.

What shell are you using?

**Comment by geeekgod** [comment]
I'm currently using zsh

and while running `./tailwindcss-linux-x64 --help` I got this as the output
```shell
zsh: exec format error: ./tailwindcss-linux-x64
```

**Comment by thecrypticace** [comment]
I'm not sure what's going on here. Whatever this is it has something to do with your install of WSL not recognizing _some_ ELF binaries which is very strange. Ultimately, I don't know that there's anything _we_ can do to fix this since I can confirm that the binary runs fine on WSL _and_ other linux x64 systems.

Do you have Discord? Any chance you could pop in our help channel and @-mention me there? https://tailwindcss.com/discord

**Comment by geeekgod** [comment]
Yes, I've discord with me, I'll ping you there!

**Comment by thecrypticace** [comment, marked as answer]
Gonna move this to discussions as it's almost certainly a environment based problem and not a problem with our binaries and the discussion is continuing in Discord until we track down the source of the problem anyway.

**Reply by geeekgod** [reply]
Yes, it was undoubtedly an environment based problem, I tried to reinstall my WSL and now can use tailwind.

**Comment by sunwooz** [comment]
Not sure if this is completed related, but I was having trouble getting the gems to build on my Render.com server, but it was working fine locally.

My fix was to delete my vendor/cache folder in the Rails root directory, which fixed my issue.

**Comment by jharrell26** [comment]
I has to switch from WSL1 to WSL2 to deal with this error message.

**Reply by xMKx** [reply]
Same here, that resolved the issue.

**Comment by igolaizola** [comment]
Anyone has figured how to fix this other than reinstalling WSL?
I have the same problem with WSL2 and Ubuntu 22.04.3 LTS.

**Reply by jharrell26** [reply]
Make sure you have the latest versions of node.js and npm. The default versions on Ubuntu 22 are not new enough.

**Reply by igolaizola** [reply]
This is happening to me with the standalone binary.

**Reply by igolaizola** [reply]
My bad... I was using the arm64 linux version instead of the amd64 one 🙈

**Comment by rapaccinim** [comment]
I was having the `The asset "tailwind.css" is not present in the asset pipeline` issue while trying to run a [Rails application with Tailwind CSS](https://tailwindcss.com/docs/guides/ruby-on-rails) on WSL.

I followed these steps:
- upgrading WSL1 to WSL2 (running in Windows PowerShell: `wsl --set-version Ubuntu 2`)
- `sudo apt` to update and upgrade the packages
- using `rvm` instead of `rbenv` because I couldn't install a more recent version of Ruby with `rbenv` (maybe I didn't explored enough other potentail solutions, so I cut short to using `rvm` as an alternative)
- `rvmsudo rvm cleanup all` to fix the `Could not extract RVM sources` issue
- selecting the right Ruby version
- finally, launching `bin/dev` I didn't have the error anymore,
